apps/functions-subscription/src/use-cases/insert-subscription.ts

```diff
@@ -0,0 +1,69 @@
+import { pipe } from 'fp-ts/lib/function';
+import * as O from 'fp-ts/Option';
+import * as TE from 'fp-ts/TaskEither';
+import * as RTE from 'fp-ts/ReaderTaskEither';
+import { Capabilities } from '../domain/capabilities';
+import {
+  SubscriptionId,
+  TrialId,
+  UserId,
+  makeSubscriptionId,
+} from '../domain/subscription';
+import { SubscriptionAlreadyExists, SubscriptionStoreError } from './errors';
+import { nowDate } from '../domain/clock';
+
+// Maps all the requirements for this use-case
+type Env = Pick<
+  Capabilities,
+  | 'subscriptionRequestWriter'
+  | 'subscriptionReader'
+  | 'subscriptionWriter'
+  | 'hashFn'
+  | 'clock'
+>;
+
+const handleSubscriptionAlreadyExists =
+  (id: SubscriptionId) => (env: Pick<Capabilities, 'subscriptionReader'>) =>
+    pipe(
+      env.subscriptionReader.get(id),
+      TE.flatMap(
+        O.fold(
+          () => TE.right(id),
+          () =>
+            TE.left(
+              new SubscriptionAlreadyExists('Subscription already exists'),
+            ),
+        ),
+      ),
+    );
+
+const handleMissingSubscription =
+  (userId: UserId, trialId: TrialId, id: SubscriptionId, now: Date) =>
+  ({ subscriptionRequestWriter, subscriptionWriter }: Env) =>
+    pipe(
+      subscriptionRequestWriter.insert({ userId, trialId }),
```

Comment:
Why not mapping left error?

Reply:
I don't think it is needed. We can differentiate the `500` error from the `429` error by changing the `subscriptionRequestWriter.insert` return type, so no mapping here is required.